File: src/main/java/de/neo/eventsystem/EventSystem.java
package de.neo.eventsystem;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;

public class EventSystem {

    private HashMap<Class<Event>, HashMap<EventPriority, HashMap<Object, Method>>> registeredListener;

    public EventSystem() {
        registeredListener = new HashMap<>();
    }

    public void registerListener(Object listenerClass) {
        for(Method method : listenerClass.getClass().getDeclaredMethods()) {
            Listener listener = method.getAnnotation(Listener.class);
            if(listener == null) {
                System.out.println(method.getName() + ": NULL");
                continue;
            }
            if(method.getParameterCount() == 1) {
                Class<Event> eventClass = (Class<Event>) method.getGenericParameterTypes()[0];
                HashMap<EventPriority, HashMap<Object, Method>> list = registeredListener.getOrDefault(eventClass,
                        new HashMap<>());
                HashMap<Object, Method> set = list.getOrDefault(listener.priority(), new HashMap<>());
                set.put(listenerClass, method);
                list.put(listener.priority(), set);
                registeredListener.put(eventClass, list);
            }
        }
    }

    public void executeEvent(Event event) {
        HashMap<EventPriority, HashMap<Object, Method>> list = registeredListener.getOrDefault(event.getClass(), new HashMap<>());
        list.getOrDefault(EventPriority.LOWEST, new HashMap<>()).forEach((k, v) -> {
            try {
                v.invoke(k, event);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        });
        list.getOrDefault(EventPriority.LOW, new HashMap<>()).forEach((k, v) -> {
            try {
                v.invoke(k, event);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        });
        list.getOrDefault(EventPriority.NORMAL, new HashMap<>()).forEach((k, v) -> {
            try {
                v.invoke(k, event);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        });
        list.getOrDefault(EventPriority.HIGH, new HashMap<>()).forEach((k, v) -> {
            try {
                v.invoke(k, event);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        });
        list.getOrDefault(EventPriority.HIGHEST, new HashMap<>()).forEach((k, v) -> {
            try {
                v.invoke(k, event);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        });
    }
}

File: src/main/kotlin/de/neo/eventsystem/kt/EventSystem.kt
package de.neo.eventsystem.kt

import java.lang.reflect.Method

class EventSystem {

    private val registeredListener = HashMap<Class<Event>, HashMap<EventPriority, HashMap<Any, Method>>>()

    fun registerListener(listenerClass: Any) {
        for(method in listenerClass.javaClass.methods) {
            if(method.annotations.isNotEmpty()) {
                val listener = method.getAnnotation(Listener::class.java)
                if(method.parameterCount == 1) {
                    val eventClass = method.parameterTypes[0] as Class<Event>
                    val list = registeredListener[eventClass] ?: HashMap()
                    val set = list[listener.priority] ?: HashMap()
                    set[listenerClass] = method
                    list[listener.priority] = set
                    registeredListener[eventClass] = list
                }
            }
        }
    }

    fun executeEvent(event: Event) {
        val list = registeredListener[event.javaClass] ?: HashMap()
        list[EventPriority.LOWEST]?.forEach { it.value.invoke(it.key, event) }
        list[EventPriority.LOW]?.forEach { it.value.invoke(it.key, event) }
        list[EventPriority.NORMAL]?.forEach { it.value.invoke(it.key, event) }
        list[EventPriority.HIGH]?.forEach { it.value.invoke(it.key, event) }
        list[EventPriority.HIGHEST]?.forEach { it.value.invoke(it.key, event) }
    }
}